app/src/main/java/org/oppia/android/app/customview/LessonThumbnailImageView.kt

```diff
@@ -69,17 +73,26 @@ class LessonThumbnailImageView @JvmOverloads constructor(
     this.isBlurred = isBlurred
   }
 
-  private fun checkIfLoadingIsPossible() {
-    if (::entityId.isInitialized &&
-      ::entityType.isInitialized &&
-      ::lessonThumbnail.isInitialized &&
-      ::thumbnailDownloadUrlTemplate.isInitialized &&
+  private val allDependenciesInjected
+    get() = ::thumbnailDownloadUrlTemplate.isInitialized &&
       ::resourceBucketName.isInitialized &&
       ::gcsPrefix.isInitialized &&
       ::imageLoader.isInitialized &&
       ::logger.isInitialized
+
+  private fun checkIfLoadingIsPossible() {
+    if (::entityId.isInitialized &&
+      ::entityType.isInitialized &&
+      ::lessonThumbnail.isInitialized
     ) {
-      loadLessonThumbnail()
+      GlobalScope.launch(Dispatchers.Default) {
```

Comment:
We should never use dispatchers other than the ones we centrally manage, otherwise it's easy to introduce performance regressions, and it will be much harder to reliably test changes since unsynchronized dispatchers can introduce flakes.

Introducing an infinite loop is also potentially verify problematic, as is the post back to the main thread. The former could cause performance issues on the user's device if it never resolves since it'll be burning 100% CPU usage for that thread, and the post back to the main thread could introduce crashes if the view is being torn down right when the push occurs (which is why we only use DataProviders + LiveData for these situations).

Per the issue described [here](https://github.com/oppia/oppia-android/issues/2780#issuecomment-796791569), could we add a call to ``checkIfLoadingIsPossible`` after inject() is called? Unless I'm missing something, that should resolve the race when databinding calls everything before injection. Or is there something else that makes this problematic?

Reply:
> Per the issue described here, could we add a call to checkIfLoadingIsPossible after inject() is called?

@BenHenning This worked. Thanks.

Should I close this draft PR and create a new one OR revert commits in this and mark it Ready for review?

Reply:
I think recycling this PR is fine. No need to revert commits--just undo the changes & add the new ones that work + tests to verify that it works (the latter is actually really important--you should always make sure to have tests that fail without your fix and pass with it when introducing a bug fix).

It's also better to keep the same PR since that retains history. We expect code to evolve during a code review--no problem with that. :)

Reply:
Okay. Will do that.
Also, can you plz tell what should be the test? 
Do I need to iterate over all items in the recycler view and check if the thumbnail has been loaded in the view or not?
Or is there some better way to test this fix?

Reply:
You'll probably want to force the situation in order to verify that this is correct. That would be introducing a LessonThumbnailImageViewTest that creates a new LessonThumbnailImageView, then calls onAttachedToWindow/the binding methods in different orders for different test cases & then verify that the thumbnail was correctly initialized in the image view. The Dagger injection setup might be a bit tricky--let me know if you run into any issues there.

If you're unsure which test cases to add, I suggest just starting with one, get that passing, and then ask for additional feedback.

Reply:
I am having some trouble figuring this out. The `onAttachedToWindow()` function finds the fragment corresponding to current view for dagger injection. Will I have to launch the activity first? And if I launch the activity, how to replace the original view with this test view?

Reply:
@ArpitShukIa just to clarify, is your question in the context of the new ``LessonThumbnailImageViewTest``? If so, you'll need both a test activity & test fragment where the test fragment can point to a view that contains a ``LessonThumbnailImageView``. This isn't much different how we set up real activities/fragments elsewhere in the codebase, or how we use ``LessonThumbnailImageView`` in real layouts (e.g. ``topic_revision_summary_view.xml``).

Reply:
The latest solution looks really solid!